rllib/utils/replay_buffers/multi_agent_episode_replay_buffer.py

```diff
@@ -0,0 +1,671 @@
+import copy
+from collections import defaultdict
+import numpy as np
+import scipy
+from typing import Dict, List, Optional, Tuple, Union
+
+from ray.rllib.core.columns import Columns
+from ray.rllib.env.multi_agent_episode import MultiAgentEpisode
+from ray.rllib.utils.replay_buffers.episode_replay_buffer import EpisodeReplayBuffer
+from ray.rllib.utils import force_list
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.spaces.space_utils import batch
+from ray.rllib.utils.typing import SampleBatchType
+
+
+class MultiAgentEpisodeReplayBuffer(EpisodeReplayBuffer):
+    def __init__(
+        self,
+        capacity: int = 10000,
+        *,
+        batch_size_B: int = 16,
+        batch_length_T: int = 64,
+        **kwargs,
+    ):
+        # Initialize the base episode replay buffer.
+        super().__init__(
+            capacity=capacity,
+            batch_size_B=batch_size_B,
+            batch_length_T=batch_length_T,
+            **kwargs,
+        )
+
+        # Stores indices of module (single-agent) timesteps.
+        self._module_indices: Dict[str, List[Tuple[int]]] = defaultdict(list)
+        # Stores for each module the episode indices.
```

Comment:
Same: improve typehint to `DefaultDict[ModuleID, Dict[EpisodeID, int]]` ? Is `int` correct here or should this be a 4-tuple as well?

Let's also above declare what "index" means throughout this entire class:
Something like:
"In this class, we use a global indexing system, where an individual index is a 4-tuple, consisting of 1) ...2)... 3)... 4) ..."